hippylib/modeling/prior.py

```diff
@@ -604,4 +605,124 @@ def sample(self, noise, s, add_mean=True):
+                                  defined. Must be the Real space with one global 
+                                  degree of freedom
+        - :code:`covariance`:     The covariance of the prior. Must be a
+                                  :code:`numpy.ndarray` of appropriate size
+        - :code:`mean`(optional): Mean of the prior distribution. Must be of
+                                  type `dolfin.Vector()`
+        """
+
+        self.Vh = Vh
+
+        if Vh.dim() != covariance.shape[0] or Vh.dim() != covariance.shape[1]:
+            raise ValueError("Covariance incompatible with Finite Element space")
+
+        self.covariance = covariance.astype(np.float64)
+        
+        #np.linalg.cholesky automatically provides error checking, so use those
+        self.chol = np.linalg.cholesky(self.covariance)
+
+        self.chol_inv = scila.solve_triangular(
+                                        self.chol,
+                                        np.identity(Vh.dim()),
+                                        lower=True)
+
+        self.precision = np.dot(self.chol_inv.T, self.chol_inv)
+
+        trial = dl.TrialFunction(Vh)
+        test  = dl.TestFunction(Vh)
+        
+        domain_measure_inv = 1.0 / dl.assemble(dl.Constant(1.) * dl.dx(Vh.mesh()))
+
+        #Identity mass matrix
+        self.M = dl.assemble(domain_measure_inv * dl.inner(trial, test) * dl.dx)
+        self.Msolver = Operator2Solver(self.M)
+
+        if mean:
+            self.mean = mean
+        else:
+            tmp = dl.Vector()
+            self.M.init_vector(tmp, 0)
+            tmp.zero()
+            self.mean = tmp
+
+        if Vh.dim() == 1:
+            trial = dl.as_matrix([[trial]])
+            test  = dl.as_matrix([[test]])
+
+        #Create form matrices 
+        covariance_form = dl.as_matrix(list(map(list, self.covariance)))
+        precision_form  = dl.as_matrix(list(map(list, self.precision)))
+        chol_form       = dl.as_matrix(list(map(list, self.chol)))
+        chol_inv_form   = dl.as_matrix(list(map(list, self.chol_inv)))
+
+        #variational for the regularization operator, or the precision matrix
+        var_form_R = domain_measure_inv \
+                     * dl.inner(test, dl.dot(precision_form, trial))
+
+        #variational for the inverse regularization operator, or the covariance
+        #matrix
+        var_form_Rinv = domain_measure_inv \
+                        * dl.inner(test, dl.dot(covariance_form, trial))
+
+        #variational form for the square root of the regularization operator
+        var_form_R_sqrt = domain_measure_inv \
+                          * dl.inner(test, dl.dot(chol_inv_form.T, trial))
+
+        #variational form for the square root of the inverse regularization 
+        #operator
+        var_form_Rinv_sqrt = domain_measure_inv \
+                             * dl.inner(test, dl.dot(chol_form, trial))
+
+        self.R         = dl.assemble(var_form_R * dl.dx)
```

Comment:
Usually in hIPPYlib, I would put the `dl.dx` with the varf not at the time of assembly. Of course it's only a matter of taste.